# Bidirectional Recurrent Neural Network (양방향 순환 신경망)

In [4]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Bidirectional, SimpleRNN

In [7]:
timesteps, input_dim, hidden_units = 10,5,8

model = Sequential()
model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True), input_shape=(timesteps, input_dim)))
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 bidirectional_1 (Bidirectio  (None, 10, 16)           224       
 nal)                                                            
                                                                 
Total params: 224
Trainable params: 224
Non-trainable params: 0
_________________________________________________________________


### `Deep Bidirenctional Recurrent Neural Network`

In [14]:
model = Sequential()
model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True), input_shape=(timesteps, input_dim)))
model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True)))
model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True)))
model.add(Bidirectional(SimpleRNN(hidden_units, return_sequences=True)))

model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 bidirectional_3 (Bidirectio  (None, 10, 16)           224       
 nal)                                                            
                                                                 
 bidirectional_4 (Bidirectio  (None, 10, 16)           400       
 nal)                                                            
                                                                 
 bidirectional_5 (Bidirectio  (None, 10, 16)           400       
 nal)                                                            
                                                                 
 bidirectional_6 (Bidirectio  (None, 10, 16)           400       
 nal)                                                            
                                                                 
Total params: 1,424
Trainable params: 1,424
Non-traina

    Q. Embedding을 사용하며, 단어 집합(Vocabulary)의 크기가 5,000이고 임베딩 벡터의 차원은 100입니다.
    은닉층에서는 Simple RNN을 사용하며, 은닉 상태의 크기는 128입니다.
    훈련에 사용하는 모든 샘플의 길이는 30으로 가정합니다.
    이진 분류를 수행하는 모델로, 출력층의 뉴런은 1개로 시그모이드 함수를 사용합니다.
    은닉층은 1개입니다.
    
    이때 RNN의 총 parameter 수?

In [21]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import SimpleRNN, Embedding, Dense


In [24]:
vocab_size = 5000
embedding_dim = 100
hidden_size=128

model = Sequential()
model.add(Embedding(vocab_size, embedding_dim))
model.add(SimpleRNN(hidden_size))
model.add(Dense(1, activation='sigmoid'))
model.summary()


Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, None, 100)         500000    
                                                                 
 simple_rnn_9 (SimpleRNN)    (None, 128)               29312     
                                                                 
 dense_2 (Dense)             (None, 1)                 129       
                                                                 
Total params: 529,441
Trainable params: 529,441
Non-trainable params: 0
_________________________________________________________________


    Embedding = 5,000(input) * 100(embedding) = 500,000
    Wx = 100(embedding) * 128(hidden) = 12,800
    Wh = 128 * 128 = 16,384
    bias(hidden)  = 128
    Wy = 128
    bias(output) = 1

    total = 529,441 